components/sup/src/manager/file_watcher.rs

```diff
@@ -1,3464 +1,2652 @@
-        self.process_args_after_settle = None;
-        self.symlink_loop_catcher.clear();
-        dirs_to_unwatch
-    }
-}
+        mod handle_remove_event {
+            use super::*;
 
-/// A regular file watcher.
-///
-/// This type watches for a regular file at any path. The file does
-/// not need to exist even - `FileWatcher` will track all the
-/// directories and symlinks from the root directory up to the
-/// file. If the file or any intermediate directory is missing, then
-/// FileWatcher will wait for it to show up.
-///
-/// `FileWatcher` will use callbacks to notify the user about events
-/// that happened to the regular file at the desired path. Note that
-/// it will call the file_appeared callback in the first iteration if
-/// the file existed when the watcher was created.
-pub struct FileWatcher<C: Callbacks, W: Watcher> {
-    callbacks:         C,
-    // The watcher itself.
-    watcher:           W,
-    // A channel for receiving events.
-    rx:                Receiver<DebouncedEvent>,
-    // The paths to watch.
-    paths:             Paths,
-    // Path to the file if it existed when we created the watcher.
-    initial_real_file: Option<PathBuf>,
-}
+            #[test]
+            fn delete_watched_file() -> Result<()> {
+                let tempdir = tempdir()?;
+                let td_enclosing_dir = tempdir.path().join("enclosing_dir");
+                fs::create_dir(&td_enclosing_dir)?;
+                let td_watched_file = td_enclosing_dir.join("watched_file.txt");
+                fs::create_file(&td_watched_file)?;
+                let tc = TestCallbacks::default();
+                let mut fw = Tfw::new(&td_watched_file, tc, true)?;
 
-/// Convenience function for returning a new file watcher that matches
-/// the platform.
-pub fn create_file_watcher<P, C>(path: P,
-                                 callbacks: C,
-                                 ignore_initial: bool)
-                                 -> Result<FileWatcher<C, SupWatcher>>
-    where P: Into<PathBuf>,
-          C: Callbacks
-{
-    FileWatcher::<C, SupWatcher>::create(path, callbacks, ignore_initial)
-}
+                fs::remove(&td_watched_file)?;
 
-impl<C: Callbacks, W: Watcher> FileWatcher<C, W> {
-    /// Creates a new `FileWatcher`.
-    ///
-    /// This will create an instance of `W` and start watching the
-    /// paths. When looping the file watcher, it will emit an initial
-    /// "file appeared" event if ignore_initial is Some(false) and the
-    /// watched file existed when the file watcher was created.
-    /// Will return `Error::NotifyCreateError` if creating the watcher
-    /// fails. In case of watching errors, it returns
-    /// `Error::NotifyError`.
-    pub fn create<P>(path: P, callbacks: C, ignore_initial: bool) -> Result<Self>
-        where P: Into<PathBuf>
-    {
-        Self::create_instance(path, callbacks, ignore_initial)
-    }
+                for _ in 0..max_test_attempts(&fw) {
```

Comment:
for removing, you don't need this loop right?

Reply:
Pretty sure I need the loop.  Once the thing that triggers the event happens, here fs:remove call, the code has to receive the event which is what happens in single_iteration.